### PR TITLE
Registration with google account

### PR DESCRIPTION
**Pull request by RezenkovD**

### CODE REVIEW

File: src/services/user.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+from sqlalchemy_utils.types import url
+
+import models
+
+
+def get_user(db: Session, login: str):
+    return db.query(models.User).filter_by(login=login).one_or_none()
+
+
+def get_users(db: Session):
```
Comment (qqlexa) [inline src/services/user.py:11]: Add return annotation.

File: src/services/user.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    return db.query(models.User).all()
+
+
+def create_user(db: Session, login: str, first_name: str, last_name: str, picture: url):
```
Comment (qqlexa) [inline src/services/user.py:15]: Add return annotation.

File: src/services/user.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+from sqlalchemy.orm import Session
+from sqlalchemy_utils.types import url
+
+import models
```
Comment (qqlexa) [inline src/services/user.py:4]: Why not ```python
from models import User
```?

**Review by qqlexa — CHANGES_REQUESTED** [review]

File: requirements.txt [hunk 4]
```diff
@@ -1,16 +1,28 @@
 anyio==3.6.2
```
[the remark this answers is not in the document]
Comment (RezenkovD) [reply requirements.txt:1]: I agree with you. This should be exclusive to the project.

File: src/routers/registration.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+from fastapi import Depends, APIRouter
+
+from database import get_db
+from services import create_user, get_user
+
+router = APIRouter(
+    tags=["registration"],
+)
+
+CONF_URL = "https://accounts.google.com/.well-known/openid-configuration"
```
[the remark this answers is not in the document]
Comment (RezenkovD) [reply src/routers/registration.py:16]: Nicely observed

File: src/routers/registration.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+async def auth(request: Request, db: Session = Depends(get_db)):
+    try:
+        token = await oauth.google.authorize_access_token(request)
+    except OAuthError as error:
+        return error.error
+    user = token["userinfo"]
+    if user:
+        request.session["user"] = dict(user)
+        db_user = get_user(db, login=user["email"])
+        if db_user:
```
[the remark this answers is not in the document]
Comment (RezenkovD) [reply src/routers/registration.py:47]: Thanks. This good idea.

File: alembic/versions/64da66deb62d_add_user_table.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+def upgrade() -> None:
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table(
+        "users",
+        sa.Column("id", sa.Integer(), nullable=False),
+        sa.Column("login", sa.String(), nullable=False),
+        sa.Column("first_name", sa.String(), nullable=False),
+        sa.Column("last_name", sa.String(), nullable=False),
+        sa.Column("picture", sqlalchemy_utils.types.url.URLType(), nullable=False),
```
[the remark this answers is not in the document]
Comment (RezenkovD) [reply alembic/versions/64da66deb62d_add_user_table.py:28]: -

File: src/routers/registration.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        "scope": "openid email profile",
+        "prompt": "select_account",
+    },
+)
+
+
+@router.get("/login")
+async def login(request: Request):
+    redirect_uri = request.url_for("auth")
+    return await oauth.google.authorize_redirect(request, redirect_uri)
```
Comment (qqlexa) [inline src/routers/registration.py:34]: Let's create tests

**Review by qqlexa — CHANGES_REQUESTED** [review]

File: src/routers/registration.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+async def auth(request: Request, db: Session = Depends(get_db)):
+    try:
+        token = await oauth.google.authorize_access_token(request)
+    except OAuthError as error:
+        return error.error
+    user = token["userinfo"]
+    request.session["user"] = dict(user)
+    db_user = get_user(db, login=user["email"])
+    if not db_user:
+        create_user(
```
Comment (qqlexa) [inline src/routers/registration.py:32]: Can we catch it from the func and later return it? db_user = create_user()

File: src/routers/registration.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+            db=db,
+            login=user["email"],
+            first_name=user["given_name"],
+            last_name=user["family_name"],
+            picture=user["picture"],
+        )
+    return get_user(db, login=user["email"])
```
Comment (qqlexa) [inline src/routers/registration.py:39]: return db_user?

File: src/services/user.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+import schemas
+
+
+def get_user(db: Session, login: str) -> Union[schemas.User, None]:
+    return db.query(models.User).filter_by(login=login).one_or_none()
+
+
+def get_users(db: Session) -> List[Type[schemas.User]]:
```
Comment (qqlexa) [inline src/services/user.py:14]: It is not `List[Type[schemas.User]]` here. It is `List[models.User]:`
By the way, what is the version of your python? It is available to use list[] from python3.9

File: src/services/user.py [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    return db.query(models.User).all()
+
+
+def create_user(db: Session, login: str, first_name: str, last_name: str, picture: url) -> schemas.User:
```
Comment (qqlexa) [inline src/services/user.py:18]: It is not `schemas.User` it is `models.User`

**Review by qqlexa — CHANGES_REQUESTED** [review]

File: src/config.py [hunk 13]
```diff
@@ -0,0 +1,13 @@
+from pydantic import BaseSettings
+
+
+class Settings(BaseSettings):
+    SQLALCHEMY_DATABASE_URI: str
+    TEST_SQLALCHEMY_DATABASE_URI: str
```
[the remark this answers is not in the document]
Comment (RezenkovD) [reply src/config.py:6]: I agree about the TEST. But if you use one variable for DB. Your main database crashes every time after unit testing.

File: src/models/user.py [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+
+
+class User(Base):
+    __tablename__ = "users"
+
+    id = Column(Integer, primary_key=True, index=True)
+    login = Column(String, unique=True, nullable=False)
+    first_name = Column(String, unique=False, nullable=False)
+    last_name = Column(String, unique=False, nullable=False)
+    picture = Column(URLType, unique=True, nullable=True)
```
[the remark this answers is not in the document]
Comment (RezenkovD) [reply src/models/user.py:14]: Exactly. After all, the picture is taken from a Google account. But you can not include it.

File: src/routers/registration.py (hunk 9 above)
Comment (RezenkovD) [reply src/routers/registration.py:32]: This good idea. Thanks.

File: src/routers/registration.py (hunk 10 above)
Comment (RezenkovD) [reply src/routers/registration.py:39]: Exclusively for testing.

File: src/services/user.py (hunk 11 above)
Comment (RezenkovD) [reply src/services/user.py:14]: 3.10 :)

File: src/services/user.py (hunk 12 above)
Comment (RezenkovD) [reply src/services/user.py:18]: Oh. Exactly. Thank you

File: src/config.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (RezenkovD) [reply src/config.py:6]: So. I understand you. There should be only one var.

File: src/services/user.py [hunk 15]
```diff
@@ -0,0 +1,22 @@
+from typing import Union, List
+
+from sqlalchemy.orm import Session
+from sqlalchemy_utils.types import url
+
+from models import User
+
+
+def get_user(db: Session, login: str) -> Union[User, None]:
```
Comment (qqlexa) [inline src/services/user.py:9]: Union[User, None] = Optional[User]

File: src/services/user.py [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+    return db.query(User).filter_by(login=login).one_or_none()
+
+
+def get_users(db: Session) -> List[User]:
+    return db.query(User).all()
+
+
+def create_user(db: Session, login: str, first_name: str, last_name: str, picture: url) -> User:
```
Comment (qqlexa) [inline src/services/user.py:17]: We use it once. Let's move into the router.

**Review by qqlexa — CHANGES_REQUESTED** [review]

**Pull request merged**